### PR TITLE
feat: add support for body factory

### PR DESCRIPTION
**Pull request by ronag**
## Status

- [ ] I have read and agreed to the [Developer's Certificate of Origin][cert]
- [ ] Tested
- [ ] Benchmarked (**optional**)
- [ ] Documented
- [ ] Review ready
- [ ] In review
- [ ] Merge ready

[cert]: https://github.com/nodejs/undici/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Comment by mcollina** [comment]
I don't understand this change, why is it needed?

**Comment by ronag** [comment]
It's for e.g enabling retries during pipelining for put request with a streaming body.

**Comment by ronag** [comment]
Ditto with redirects

**Comment by ronag** [comment]
You pass a factory method that can always re create the stream body from 0.

**Comment by ronag** [comment]
Maybe a better example:

```js
request(url, {
  method: 'GET',
  body: () => fs.createReadStream(path)
})
```

File: types/dispatcher.d.ts [hunk 1]
```diff
@@ -98,7 +98,7 @@ declare namespace Dispatcher {
     path: string;
     method: HttpMethod;
     /** Default: `null` */
-    body?: string | Buffer | Uint8Array | Readable | null | FormData;
+    body?: string | Buffer | Uint8Array | Readable | null | FormData | Function;
```
Comment (metcoder95) [inline types/dispatcher.d.ts:101]: I believe we would need to document better this 🤔

File: lib/client.js [hunk 2]
```diff
@@ -1472,12 +1474,59 @@ function shouldSendContentLength (method) {
 }
 
 function write (client, request) {
+  const socket = client[kSocket]
+  const { body } = request
+
+  if (typeof body === 'function') {
+    // TODO (fix): What if socket is destroyed while
+    // waiting for body?
```
Comment (metcoder95) [inline lib/client.js:1482]: If the factory is meant to be idempotent, destroying the body should be ok to do too.
We should advice to be careful about the state of the body

**Comment by mcollina** [comment]
This seems something I would add to the `dispatcher.stream()` method.

https://github.com/nodejs/undici/blob/main/docs/api/Dispatcher.md#dispatcherstreamoptions-factory-callback

**Comment by ronag** [comment]
Sorry. But I'm not sure how that's relevant here?

**Comment by mcollina** [comment]
This kind of API seemed nice there.

**Comment by ronag** [comment]
Sorry but I still don't understand. This cannot be implemented there. For the pipeline retry to work this has to live in the client. Putting it in stream doesn't do anything?

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request closed without merging**